pkg/prometheus/ha/codec.go

```diff
@@ -60,3 +68,45 @@ func (*yamlCodec) Encode(v interface{}) ([]byte, error) {
 func (*yamlCodec) CodecID() string {
 	return "agentConfig/yaml"
 }
+
+type codecConfig struct {
+	Config  *instance.Config `yaml:"config"`
+	Secrets []configSecrets  `yaml:"secrets"`
+}
+
+// newCodecConfig creates a new codecConfig that is ready for marshaling as YAML
+// and storing in a KV store. newCodecConfig must extract secrets stored in the
+// instance.Config and store them elsewhere as strings as the MarshalYAML function
+// on the Prometheus Secret type replaces the contents of the secret with
+// the text "<secret>".
+func newCodecConfig(c *instance.Config) *codecConfig {
+	var secrets []configSecrets
+	for _, rwr := range c.RemoteWrite {
```

Comment:
is order when marshalling/unmarshalling guaranteed to be preserved?

Reply:
Also, this means that if you move sections in the file it is now broken? Maybe this is isn't a problem if we don't support it anyways.

Reply:
Yep, the [JSON spec](https://www.rfc-editor.org/rfc/rfc7159.txt) says that arrays are ordered - this seems to hold in our [YAML decoder](https://github.com/go-yaml/yaml/blob/0b1645d91e851e735d3e23330303ce81f70adbe3/decode.go#L201) as well.

As for moving sections, we require users to always upload the whole config file each time (including the secrets). So if you move the sections around in the file and update the config in the API, `newCodecConfig` with the write ordering would completely supplant the old one. 

It's possible someone could modify the KV store directly and mess up our ordering, but there's probably nothing we can (or should?) do about that :)

Reply:
Makes perfect sense.